pkg/controller/apischeme/apischeme_controller.go

```diff
@@ -96,5 +119,320 @@ func (r *ReconcileAPIScheme) Reconcile(request reconcile.Request) (reconcile.Res
+	}
+	masterNodeInstances, err := utils.GetClusterMasterInstances(r.client)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't detect the AWS instances for master nodes")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't find the cluster's AWS instances for master nodes", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+
+	// In theory this could return more than one VPC for master nodes. That would be
+	// odd since the security group is associated with a single VPC.
+	vpcs, err := utils.GetMasterNodeVPCs(r.client)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't get the VPC in use by master nodes")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't get the VPC id for master nodes", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+	if len(vpcs) > 1 {
+		reqLogger.Info(fmt.Sprintf("Multiple VPCs detected for master nodes. Using %s for security group", vpcs[0]))
+	}
+
+	lb := &LoadBalancer{
+		EndpointName:     instance.Spec.ManagementAPIServerIngress.DNSName,
+		Subnets:          subnets,
+		Tags:             ownerTags,
+		MachineInstances: masterNodeInstances,
+		BaseDomain:       clusterBaseDomain,
+	}
+	cidrAccess := &CIDRAccess{
+		LoadBalancer:         lb,
+		SecurityGroupName:    config.AdminAPISecurityGroupName,
+		SecurityGroupVPCName: vpcs[0],
+		Tags:                 ownerTags,
+	}
+
+	// Now try to make sure all the things for Admin API are present in AWS
+	err = ensureAdminAPIEndpoint(reqLogger, instance, awsClient, lb, cidrAccess)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't ensure the admin API endpoint")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't ensure the admin API endpoint: "+err.Error(), cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+
+	// And now, tell the APIServer/cluster object about it.
+	err = r.addAdminAPIToAPIServerObject(reqLogger, instance.Spec.ManagementAPIServerIngress.DNSName+"."+clusterBaseDomain)
+	if err != nil {
+		reqLogger.Error(err, "Couldn't update APIServer/cluster object")
+		SetAPISchemeStatus(reqLogger, instance, "Couldn't reconcile", "Couldn't update APIServer/cluster object", cloudingressv1alpha1.ConditionError)
+		r.client.Status().Update(context.TODO(), instance)
+		return reconcile.Result{}, err
+	}
+	SetAPISchemeStatus(reqLogger, instance, "Success", "Admin API Endpoint created", cloudingressv1alpha1.ConditionReady)
+	r.client.Status().Update(context.TODO(), instance)
+
+	return reconcile.Result{}, nil
+}
+
+func ensureCloudLoadBalancer(reqLogger logr.Logger, awsAPI *awsclient.AwsClient, lb *LoadBalancer) (*awsclient.AWSLoadBalancer, error) {
+	var awsObj *awsclient.AWSLoadBalancer
+	found := false
+	var err error
+	for i := 1; i <= config.MaxAPIRetries; i++ {
```

Comment:
Why do we set MaxAPIEntries = 10? just curious

Reply:
I picked it out of the air. No good reason for it.